label/middleware_auth.go

```diff
@@ -0,0 +1,145 @@
+package label
+
+import (
+	"context"
+	"errors"
+
+	"github.com/influxdata/influxdb/v2"
+	"github.com/influxdata/influxdb/v2/authorizer"
+)
+
+var _ influxdb.LabelService = (*AuthedLabelService)(nil)
+
+type labelContext string
+
+const ctxOrgKey labelContext = "orgID"
+
+type AuthedLabelService struct {
+	s influxdb.LabelService
+}
+
+// NewAuthedLabelService constructs an instance of an authorizing label serivce.
+func NewAuthedLabelService(s influxdb.LabelService) *AuthedLabelService {
+	return &AuthedLabelService{
+		s: s,
+	}
+}
+func (s *AuthedLabelService) CreateLabel(ctx context.Context, l *influxdb.Label) error {
+	if _, _, err := authorizer.AuthorizeCreate(ctx, influxdb.LabelsResourceType, l.OrgID); err != nil {
+		return err
+	}
+	return s.s.CreateLabel(ctx, l)
+}
+
+func (s *AuthedLabelService) FindLabels(ctx context.Context, filter influxdb.LabelFilter, opt ...influxdb.FindOptions) ([]*influxdb.Label, error) {
+	// TODO: we'll likely want to push this operation into the database eventually since fetching the whole list of data
+	// will likely be expensive.
+	ls, err := s.s.FindLabels(ctx, filter, opt...)
+	if err != nil {
+		return nil, err
+	}
+	ls, _, err = authorizer.AuthorizeFindLabels(ctx, ls)
+	return ls, err
+}
+
+// FindLabelByID checks to see if the authorizer on context has read access to the label id provided.
+func (s *AuthedLabelService) FindLabelByID(ctx context.Context, id influxdb.ID) (*influxdb.Label, error) {
+	l, err := s.s.FindLabelByID(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	if _, _, err := authorizer.AuthorizeRead(ctx, influxdb.LabelsResourceType, id, l.OrgID); err != nil {
+		return nil, err
+	}
+	return l, nil
+}
+
+// FindResourceLabels retrieves all labels belonging to the filtering resource if the authorizer on context has read access to it.
+// Then it filters the list down to only the labels that are authorized.
+func (s *AuthedLabelService) FindResourceLabels(ctx context.Context, filter influxdb.LabelMappingFilter) ([]*influxdb.Label, error) {
+	if err := filter.ResourceType.Valid(); err != nil {
+		return nil, err
+	}
+	// first fetch all labels for this resource
+	ls, err := s.s.FindResourceLabels(ctx, filter)
+	if err != nil {
+		return nil, err
+	}
+
+	// check the permissions for the resource by the org on the context
+	orgID := orgIDFromContext(ctx)
```

Comment:
@GeorgeMac the original authorizer code passed in an Org Service and got the OrgID by looking up the resource on the filter, I changed it to use the OrgID on the context to make things a little cleaner and more separated. Will this result in unwanted behavior? (It does pass the existing tests if I put the same OrgID on the context but there could be something I'm missing.)

Reply:
Oooh interesting. My only concern is my lack of understanding around what puts this orgID on the context. But if it is consistent perhaps it doesn't matter. Will think about it through this review :D This could be all we need.

Reply:
wound up making `OrgIDFromContext` and `ValidResource` public and moving them from tenant into kit/transport/http. Labels is always nested after all so the org ID being added to the context by `ValidResource` when mounting the routes in the tenant service should be sufficient